cmd/gen-release-notes/main.go

```diff
@@ -110,19 +114,198 @@ func main() {
-		"prevMilestone":    prevMilestone,
-		"changeLogSince":   changeLogSince,
-		"content":          content,
-		"k8sVersion":       k8sVersion,
-		"changeLogVersion": markdownVersion,
-		"majorMinor":       majorMinor,
+		"milestone":                   milestone,
+		"prevMilestone":               prevMilestone,
+		"changeLogSince":              changeLogSince,
+		"content":                     content,
+		"k8sVersion":                  k8sVersion,
+		"changeLogVersion":            markdownVersion,
+		"majorMinor":                  majorMinor,
+		"EtcdVersion":                 getGoModVersion("etcd", repo, milestone),
+		"ContainerdVersion":           getGoModVersion("containerd", repo, milestone),
+		"RuncVersion":                 getGoModVersion("runc", repo, milestone),
+		"CNIPluginsVersion":           getImageTagVersion("cni-plugins", repo, milestone),
+		"MetricsServerVersion":        getImageTagVersion("metrics-server", repo, milestone),
+		"TraefikVersion":              getImageTagVersion("traefik", repo, milestone),
+		"CoreDNSVersion":              getImageTagVersion("coredns", repo, milestone),
+		"IngressNginxVersion":         getDockerfileVersion("rke2-ingress-nginx", repo, milestone),
+		"HelmControllerVersion":       getGoModVersion("helm-controller", repo, prevMilestone),
+		"FlannelVersionRKE2":          getImageTagVersion("flannel", repo, milestone),
+		"FlannelVersionK3S":           getGoModVersion("flannel", repo, milestone),
+		"CalicoVersion":               calicoVersion,
+		"CalicoVersionTrimmed":        calicoVersionTrimmed,
+		"CiliumVersion":               getImageTagVersion("cilium-cilium", repo, milestone),
+		"MultusVersion":               getImageTagVersion("multus-cni", repo, milestone),
+		"KineVersion":                 getGoModVersion("kine", repo, milestone),
+		"SQLiteVersion":               sqliteVersionBinding,
+		"SQLiteVersionReplaced":       strings.ReplaceAll(sqliteVersionBinding, ".", "_"),
+		"LocalPathProvisionerVersion": getImageTagVersion("local-path-provisioner", repo, milestone),
 	}); err != nil {
 		fmt.Println(err)
 		os.Exit(1)
 	}
 
 	os.Exit(0)
 }
+
+func getGoModVersion(libraryName, repo, branchVersion string) string {
+	repoName := "k3s-io/k3s"
+	if repo == "rke2" {
+		repoName = "rancher/rke2"
+	}
+	goModURL := "https://raw.githubusercontent.com/" + repoName + "/" + branchVersion + "/go.mod"
+	resp, err := http.Get(goModURL)
+	if err != nil {
+		fmt.Printf("failed to fetch go.mod file from %s: %v\n", goModURL, err)
+		os.Exit(1)
+	}
+	if resp.StatusCode != http.StatusOK {
+		fmt.Printf("status error: %v\n", resp.StatusCode)
+		os.Exit(1)
+	}
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		fmt.Printf("read body error: %v\n", err)
+		os.Exit(1)
+	}
+	defer resp.Body.Close()
+
+	return findLibraryVersion(string(data), libraryName)
+}
+
+func findLibraryVersion(goModStr, libraryName string) string {
+	scanner := bufio.NewScanner(strings.NewReader(goModStr))
```

Comment:
Consider using https://pkg.go.dev/golang.org/x/mod/modfile#File instead of just string searching/splitting

Reply:
I grouped the redundant code so that it will use the same function for fetching/finding a word in a url based on a regex, so I dont think we would need a special code for the mod file